### PR TITLE
Simple gradient descent works for callables without gradient

### PR DESCRIPTION
**Pull request by VolodyaCO**
## Description

Previously, the simple gradient descent optimiser accepted both callables and callables with gradient. However, when passing a callable, it would not compute gradients with a numerical approximation routine, and the optimisation would fail. This PR wraps the incoming cost function, if it's just a callable, to convert it to a callable with gradient by using finite differences.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

File: src/orquestra/opt/optimizers/simple_gradient_descent.py [hunk 1]
```diff
@@ -62,22 +67,55 @@ def _minimize(
                 see note.
 
         """
-        assert isinstance(cost_function, CallableWithGradient)
-
+        # So that mypy does not complain about missing attributes:
+        assert hasattr(cost_function, "gradient")
```
Comment (AthenaCaesura) [inline src/orquestra/opt/optimizers/simple_gradient_descent.py:71]: Hmmm mypy shouldn't complain here. Might be worth talking to @dexter2206 about this.

File: tests/orquestra/opt/optimizers/simple_gradient_descent_test.py [hunk 2]
```diff
@@ -62,14 +62,13 @@ def test_fails_to_initialize_when_number_of_iterations_is_negative(self):
         with pytest.raises(AssertionError):
             SimpleGradientDescent(0.1, -1)
 
-    def test_fails_to_minimize_when_cost_function_does_not_have_gradient_method(
+    def test_succeeds_to_minimize_when_cost_function_does_not_have_gradient_method(
```
Comment (AthenaCaesura) [inline tests/orquestra/opt/optimizers/simple_gradient_descent_test.py:65]:
```suggestion
    def test_minimize_succeeds_when_cost_function_does_not_have_gradient_method(
```
Just a small chang here for clarity.

File: tests/orquestra/opt/optimizers/simple_gradient_descent_test.py [hunk 3]
```diff
@@ -62,14 +62,13 @@ def test_fails_to_initialize_when_number_of_iterations_is_negative(self):
...
         self, optimizer
     ):
         def cost_function(x):
             return sum(x)
 
-        with pytest.raises(AssertionError):
-            optimizer.minimize(cost_function, np.array([0, 0]))
+        optimizer.minimize(cost_function, np.array([0, 0]))
```
Comment (AthenaCaesura) [inline tests/orquestra/opt/optimizers/simple_gradient_descent_test.py:71]: Is it possible to put some assertion on the output of `minimize`?

File: src/orquestra/opt/optimizers/simple_gradient_descent.py (hunk 1 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/simple_gradient_descent.py:71]: So maybe it's just my linter? I think I'm using pylance or something like that.

File: tests/orquestra/opt/optimizers/simple_gradient_descent_test.py (hunk 3 above)
Comment (VolodyaCO) [reply tests/orquestra/opt/optimizers/simple_gradient_descent_test.py]: Yes, I am now checking both the parameters and the value.

**Review by AthenaCaesura — APPROVED** [review]

**Pull request merged**